hedera-node/src/main/java/com/hedera/services/txns/crypto/CryptoApproveAllowanceTransitionLogic.java

```diff
@@ -107,100 +112,142 @@ public void doStateTransition() {
 		final var op = cryptoAllowanceTxn.getCryptoApproveAllowance();
-		final var ownerAccount = accountStore.loadAccount(Id.fromGrpcAccount(owner));
+		final var payerAccount = accountStore.loadAccount(Id.fromGrpcAccount(payer));
 
-		return allowanceChecks.allowancesValidation(op.getCryptoAllowancesList(),
-				op.getTokenAllowancesList(), op.getNftAllowancesList(), ownerAccount,
+		return allowanceChecks.allowancesValidation(
+				op.getCryptoAllowancesList(),
+				op.getTokenAllowancesList(),
+				op.getNftAllowancesList(),
+				payerAccount,
 				dynamicProperties.maxAllowanceLimitPerTransaction());
 	}
 
 	/**
 	 * Applies all changes needed for Crypto allowances from the transaction
 	 *
 	 * @param cryptoAllowances
-	 * @param ownerAccount
+	 * @param payerAccount
 	 */
-	private void applyCryptoAllowances(final List<CryptoAllowance> cryptoAllowances, final Account ownerAccount) {
+	private void applyCryptoAllowances(final List<CryptoAllowance> cryptoAllowances, final Account payerAccount) {
 		if (cryptoAllowances.isEmpty()) {
 			return;
 		}
-		final Map<EntityNum, Long> cryptoMap = new TreeMap<>(ownerAccount.getCryptoAllowances());
 		for (final var allowance : cryptoAllowances) {
+			final var owner = allowance.getOwner();
+			final var accountToApprove = fetchOwnerAccount(owner, payerAccount, accountStore, entitiesChanged);
+			final Map<EntityNum, Long> cryptoMap = new TreeMap<>(accountToApprove.getCryptoAllowances());
+
 			final var spender = Id.fromGrpcAccount(allowance.getSpender());
+			accountStore.loadAccountOrFailWith(spender, INVALID_ALLOWANCE_SPENDER_ID);
+
 			final var amount = allowance.getAmount();
 			if (cryptoMap.containsKey(spender.asEntityNum())) {
 				if (amount == 0) {
-					cryptoMap.remove(spender.asEntityNum());
+					removeEntity(cryptoMap, spender, accountToApprove);
 				}
 				// No-Op need to submit adjustAllowance to adjust any allowances
 				continue;
 			}
 			cryptoMap.put(spender.asEntityNum(), amount);
+			accountToApprove.setCryptoAllowances(cryptoMap);
+			entitiesChanged.put(accountToApprove.getId().num(), accountToApprove);
+			validateFalse(exceedsAccountLimit(accountToApprove), MAX_ALLOWANCES_EXCEEDED);
 		}
-		ownerAccount.setCryptoAllowances(cryptoMap);
+	}
+
+	private void removeEntity(final Map<EntityNum, Long> cryptoMap,
+			final Id spender,
+			final Account accountToApprove) {
+		cryptoMap.remove(spender.asEntityNum());
+		accountToApprove.setCryptoAllowances(cryptoMap);
+		entitiesChanged.put(accountToApprove.getId().num(), accountToApprove);
 	}
 
 	/**
 	 * Applies all changes needed for NFT allowances from the transaction
 	 *
 	 * @param nftAllowances
-	 * @param ownerAccount
+	 * @param payerAccount
 	 */
-	private void applyNftAllowances(final List<NftAllowance> nftAllowances, final Account ownerAccount) {
+	private void applyNftAllowances(final List<NftAllowance> nftAllowances, final Account payerAccount) {
```

Comment:
Is there a place where we de-dup the NFT serials? Don't see it here.

Reply:
@superboo  Do you mean if we validate duplicate serial numbers ?
It is done in precheck [here](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/txns/crypto/validators/ApproveAllowanceChecks.java#L199) for Approve and [here](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/txns/crypto/validators/AdjustAllowanceChecks.java#L219) for Adjust

Reply:
thx